Azure/azure-cli
# [Compute] New command group sshkey

**Description**

New command group `sshkey`
Integrated in `vm create`

### SSH verb
APIs defined in Swagger:
1. "/subscriptions/{subscriptionId}/providers/Microsoft.Compute/sshPublicKeys"
GET: Lists all of the SSH public keys in the subscription.
2. "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Compute/sshPublicKeys"
GET: Lists all of the SSH public keys in the specified resource group.
3. "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Compute/sshPublicKeys/{sshPublicKeyName}"
PUT: Creates a new SSH public key resource.
PATCH: Updates a new SSH public key resource.
DELETE: Delete an SSH public key.
GET: Retrieves information about an SSH public key.
4. "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Compute/sshPublicKeys/{sshPublicKeyName}/generateKeyPair"
POST: Generates and returns a public/private key pair and populates the SSH public key resource with the public key. The length of the key will be 3072 bits. This operation can only be performed once per SSH public key resource.

**Create a key**
```
az sshkey create -g RG -n NAME --public-key VALUE
az sshkey create -g RG -n NAME
```
If --public-key is not provided, it will create a key with a value that is auto-generated using generateKeyPair API.
**Update a key**
```
az sshkey update -g RG -n NAME --public-key VALUE
```
Update a key value.
**Other operations**
Other operations are trivial.

### VM create
**Use an existing key**
```
az vm create … --ssh-key-name NAME
```
The key must exist. Otherwise, an error will be thrown. If they key does not exist an error is returned advising the user to add either `--ssh-key-value` or `--generate-ssh-keys`. 
We must be aware of a fact that if no --ssh-key-values and --generate-ssh-keys are provided, CLI will look for key in .ssh folder. If --ssh-key-name is provided, this key will be used instead of key in .ssh folder.
**Create and use new key**
```
az vm create --ssh-key-name NAME --generate-ssh-keys
```
Create a new key resource with name NAME and generated value then use this value to create the VM. NAME should not exist. Otherwise, an error will be thrown.


**Testing Guide**

Testing code
```
        self.kwargs.update({
            'key': 'ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABgQC+MQ7LmEsaB7e/H63lCxJzrWdaLVhuUsnwXD5Eo7QpNhG6g9Oj9fKmZTSdHRpnUdVGtyRUJbmEHoeqFBBAt8bHu2bneEJeh8qfRmj0lCJA2QTZsdGlCsVlfSQzMjv/2WiOZ07pPGFVKvwsNS3dYQ1LtsNDAT4KE7ITlCcNjc+BjfWFTYOplAO++nruv+mD8zF1wwgTln/tHs7Ieja9Noon4PqnvyTYExPx7pclDjIPC+FzBrd9JBk+IUZyYPETO5F/LWh0M/+R656SCvHnXZ+xgan+V6nFJ0mGMErUrXUYMyYp8n/k5G5uxAiHjbS6b/+7HGbGLC0OUCBXLB0UyfIBo5ZtOgH9JKbRd2u7hjPBza7SY52JUsHbt7gZU46W35WjbDnW+clB+qLArHrsGr3YvkrEFn0IaD8y/7O9DW0PJFHM8iFZdZqmT+zM/BFse+p9El08MjPydTfKrZW4fzSBogI4oxY42CRDzxTl/WbpuGkjfcGfKwSoDbSy9jqjD/0='
        })
        self.cmd('sshkey create -g {rg} -n k1')
        self.cmd('sshkey list')
        self.cmd('sshkey show -g {rg} -n k1', checks=[
            self.check('name', 'k1'),
            self.exists('publicKey')
        ])
        self.cmd('sshkey create -g {rg} -n k2 --public-key "{key}"')

        # Use existing key
        self.cmd('vm create -g {rg} -n vm1 --image centos --nsg-rule None --ssh-key-name k1')

        # Create new one
        self.cmd('vm create -g {rg} -n vm3 --image centos --nsg-rule None --ssh-key-name k3 --generate-ssh-keys')
        self.cmd('sshkey show -g {rg} -n k3')
```


**History Notes**


[Component Name 1] BREAKING CHANGE: az command a: Make some customer-facing breaking change.
[Component Name 2] az command b: Add some customer-facing feature.

---

This checklist is used to make sure that common guidelines for a pull request are followed.
